optimum/neuron/utils/cache_utils.py

```diff
@@ -127,7 +129,59 @@ def is_private_repo(repo_id: str) -> bool:
     return private
 
 
+_CACHED_HAS_WRITE_ACCESS_TO_REPO = {}
+
+
+def cache_has_write_access(user_function):
+    """
+    Makes it possible to cache the results of functions that check if the user has write access to a given repo while
+    handling the case where tokens were updated.
+    This is useful to not request the HF Hub too much when it is not needed.
+    """
+
+    @functools.wraps(user_function)
+    def wrapper(repo_id: str):
+        token = get_token()
+        key = (token, repo_id)
+        if key in _CACHED_HAS_WRITE_ACCESS_TO_REPO:
+            return _CACHED_HAS_WRITE_ACCESS_TO_REPO[key]
+        result = user_function(repo_id)
+        _CACHED_HAS_WRITE_ACCESS_TO_REPO[key] = result
+        return result
+
+    return wrapper
+
+
+@cache_has_write_access
 def has_write_access_to_repo(repo_id: str) -> bool:
+    api = HfApi()
+    try:
+        api.delete_branch(repo_id=repo_id, repo_type="model", branch=f"this-branch-does-not-exist-{uuid4()}")
+    except GatedRepoError:
+        return False
+    except RepositoryNotFoundError:
+        # We could raise an error to indicate the user that the repository could not even be found:
+        # raise ValueError(f"Repository {repo_id} not found (repo_type: {repo_type}). Is it a private one?") from e
+        # But here we simply return `False`, because it means that we do not have write access to this repo in the end.
+        return False
+    except RevisionNotFoundError:
+        return True  # has write access, otherwise would have been 403 forbidden.
+    except HfHubHTTPError as e:
+        if e.response.status_code == 403:
+            return False
+        else:
+            raise ValueError(f"Cannot determine write access to {repo_id}") from e
+
+
+@cache_has_write_access
+def has_write_access_to_repo_without_api_action(repo_id: str) -> bool:
```

Comment:
where do you plan to use it?

Reply:
I actually do not plan to use it, it is more like if we need that in the future. We can remove that as well.

Reply:
Removed it!